accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,951 @@
+            such that `Iterable<T>` is a superinterface of `S` (the result of
+            constraint matching for `X` using the constraint `S <:
+            Iterable<X>`).
+
+        *   If `S` is a non-`Null` subtype of `Map<Object, Object>`, then the
+            inferred key type of `element` is `K` and the inferred value type of
+            `element` is `V`, where `K` and `V` are the types such that `Map<K,
+            V>` is a superinterface of `S` (the result of constraint matching
+            for `X` and `Y` using the constraint `S <: Map<X, Y>`).
+
+            *Note that both this and the previous case can match on the same
+            element if `S` is a subtype of both `Iterable<Object>` and
+            `Map<Object, Object>`. In that case, we rely on other elements to
+            disambiguate.*
+
+        *   If `S` is `dynamic`, then the inferred set element type of `element`
+            is `dynamic`, the inferred key type of `element` is `dynamic`, and
+            the inferred value type of `element` is `dynamic`. *(We produce both
+            a set element type here, and a key/value pair here and rely on other
+            elements to disambiguate.)*
+
+        *   If `S` is `Null` and the spread operator is `...?` then the element
+            has set element type `Null`, map key type `Null` and map value type
+            `Null`.
+
+        *   If none of these cases match, it is an error.
+
+    *   If `P` is `Set<Ps>` then let `S` be the inferred type of `e1` in context
+        `Iterable<Ps>`:
+
+        *   If `S` is a non-`Null` subtype of `Iterable<Object>`, then the
+            inferred set element type of `element` is `T` where `T` is the type
+            such that `Iterable<T>` is a superinterface of `S` (the result of
+            constraint matching for `X` using the constraint `S <:
+            Iterable<X>`).
+
+        *   If `S` is `dynamic`, then the inferred set element type of `element`
+            is `dynamic`.
+
+        *   If `S` is `Null` and the spread operator is `...?`, then the set
+            element type is `Null`.
+
+        *   Otherwise it is an error.
+
+    *   If `P` is `Map<Pk, Pv>` then let `S` be the inferred type of `e1` in
+        context `P`:
+
+        *   If `S` is a non-`Null` subtype of `Map<Object, Object>`, then the
+            inferred key type of `element` is `K` and the inferred value type of
+            `element` is `V`, where `K` and `V` are the types such that `Map<K,
+            V>` is a superinterface of `S` (the result of constraint matching
+            for `X` and `Y` using the constraint `S <: Map<X, Y>`).
+
+        *   If `S` is `dynamic`, then the inferred key type of `element` is
+            `dynamic`, and the inferred value type of `element` is `dynamic`.
+
+        *   If `S` is `Null` and the spread operator is `...?`, then the key and
+            value element types are `Null`.
+
+        *   Otherwise it is an error.
```

Comment:
I'm guessing that it's also an error if `P` is neither `?`, `Set<Ps>` nor `Map<Pk, Pv>`. (In other words, I think line 276 needs to be duplicated except with one less level of indentation.)

Reply:
There's a level of indirection in the spec that makes it a little less obvious what's going on here. `P` isn't just the straight downwards inference context type from the surrounding expression. Instead, the previous section ("In a `setOrMapLiteral` *collection*, the inferred type...") defines explicitly how the original downwards inference context type gets mapped to `P`. In that section, the only types `P` may end up having are either `Set`, `Map`, or `?`.